### PR TITLE
Deprecate VPA

### LINKED ISSUE #1790: Remove Verticalpodautoscaler Collector (VPA)
As we have now an implementation to collect metrics from Custom Resources,  we want to remove the VPA Collector in KSM v3.0

### PR DESCRIPTION
**Pull request by rexagod**
- [x] Deprecate VPA metrics.
- [x] Inform users about the deprecation (print a warning to `stdout`).
- [x] Document an alternative approach using the CRD capabilities we have now.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

**How does this change affect the cardinality of KSM**: No change (VPA metrics will be removed in v2.9.0).

**Which issue(s) this PR fixes**: Partially targets #1718.

### CODE REVIEW

**Review by fpetkovski — APPROVED** [review]
Thanks.

/lgtm
/hold

File: internal/store/builder.go
Comment (dgrisonnet) [inline internal/store/builder.go:100]: Instead of adding the deprecation notice here, we should reuse the existing mechanism in the family generators: https://github.com/kubernetes/kube-state-metrics/blob/master/pkg/metric_generator/generator.go#L30-L37

File: docs/verticalpodautoscaler-metrics.md
Comment (dgrisonnet) [inline docs/verticalpodautoscaler-metrics.md:20]: Could we provide them with a manifest that generates exactly the same metrics as we had before? I know that we won't be able to get the exact same names so it would get great to also point the name changes.

File: docs/verticalpodautoscaler-metrics.md
Comment (dgrisonnet) [inline docs/verticalpodautoscaler-metrics.md:5]: Since v2.6.0 has already been released, we should only start removing VPA in v2.9.0 since the deprecation period should be of at least 2 releases.

Comment (mrueg) [reply docs/verticalpodautoscaler-metrics.md]: IMHO, we should bump version to v3 if we remove anything.

**Comment by dgrisonnet** [comment]
The unit tests are failing, you need to the update the test metadata to include the deprecation notice: https://github.com/kubernetes/kube-state-metrics/blob/master/internal/store/verticalpodautoscaler_test.go#L33-L40

File: docs/verticalpodautoscaler-metrics.md
Comment (dgrisonnet) [reply docs/verticalpodautoscaler-metrics.md]: Breaking changes used to be reserved for major releases when following semantic versioning, but what we've seen in recent years is that projects with only minor releases are healthy and can still handle breaking changes properly as long as they have clear deprecation policies, which is the case for Kubernetes for example. IMHO major releases should only be reserved for major rewrites, which don't include VPA metrics removal. If we are not confident that the deprecation period is enough, I am fine with increasing it, but I don't think waiting for a major release is the right call here.

What do you think @fpetkovski?

Comment (mrueg) [reply docs/verticalpodautoscaler-metrics.md]: I can understand that major releases can become quite painful with go versioning. Personally, I prefer going with the least surprise for users. Users won't be happy when their VPA metrics are gone and they did not see it coming.

**Comment by rexagod** [comment]
@dgrisonnet @mrueg Moved to v3, PTAL.

**Comment by rexagod** [comment]
Moved to v3, ready for review.

**Comment by rexagod** [comment]
cc @dgrisonnet @mrueg @logicalhan

File: docs/verticalpodautoscaler-metrics.md
Comment (logicalhan) [inline docs/verticalpodautoscaler-metrics.md:33]: is this output still correct given your change with the CRD metrics in your other PR? Shouldn't this be `kube_cred_annotations`?

File: docs/verticalpodautoscaler-metrics.md
Comment (dgrisonnet) [reply docs/verticalpodautoscaler-metrics.md]: The problem is that there is no reason to cut a new major today besides removing the VPA metrics, which is not a big reason to have a new major version just for. Leaving a long enough deprecation period should already be good enough to not surprise the users. At the very least if we are not confident enough that our users are going through our changelogs, we can add a log line when the vpa store is used saying that it is deprecated and will be removed in version x.y.z.

**Comment by dgrisonnet** [comment]
I am still very much against cutting a new major version of ksm just for deprecating VPA metrics. This is not an overhaul of the codebase, we are just deprecating some metrics and replacing them with new ones. See it as a renaming of metrics rather than a complete removal.

Also, it is worth noting that all the VPA metrics are experimental so although they have been in the codebase for a while, users shouldn't expect them to never change.

Holding until we've reached a consensus.
/hold

**Comment by logicalhan** [comment]
> I am still very much against cutting a new major version of ksm just for deprecating VPA metrics. This is not an overhaul of the codebase, we are just deprecating some metrics and replacing them with new ones. See it as a renaming of metrics rather than a complete removal.
> 
> Also, it is worth noting that all the VPA metrics are experimental so although they have been in the codebase for a while, users shouldn't expect them to never change.
> 
> Holding until we've reached a consensus. /hold

Yeah, cutting a release for deleting experimental metrics seems a bit much to me too.

File: docs/verticalpodautoscaler-metrics.md
Comment (logicalhan) [reply docs/verticalpodautoscaler-metrics.md]: Experimental features should not have stability guarantees.

Comment (mrueg) [reply docs/verticalpodautoscaler-metrics.md]: Apologies here, I wasn't aware that VPA never made it stable. I'm fine with deprecating and removing without bumping major here. Thanks for explaining!

**Comment by rexagod** [comment]
Reverted to `v2.9.0`.

File: docs/verticalpodautoscaler-metrics.md
Comment (mrueg) [inline docs/verticalpodautoscaler-metrics.md:27]: Should this should this say "Deprecated since v2.7.0, will be removed in v2.9.0" instead?

**Comment by rexagod** [comment]
@mrueg Can you please `/lgtm` (if all looks good here)?

**Comment by mrueg** [comment]
/lgtm
/hold cancel
Proceeding here, thanks for your contribution @rexagod!

**Comment by dgrisonnet** [comment]
There is a doc explaining how to get them back in https://github.com/kubernetes/kube-state-metrics/blob/main/docs/customresourcestate-metrics.md#verticalpodautoscaler

**Pull request merged**
